masonry/src/passes/update.rs

```diff
@@ -620,16 +623,9 @@ fn update_focus_chain_for_widget(
     state.item.has_focus = global_state.focused_widget == Some(id);
     let had_focus = state.item.has_focus;
 
-    state.item.in_focus_chain = false;
     state.item.focus_chain.clear();
-    {
-        let mut ctx = LifeCycleCtx {
-            global_state,
-            widget_state: state.item,
-            widget_state_children: state.children.reborrow_mut(),
-            widget_children: widget.children.reborrow_mut(),
-        };
-        widget.item.lifecycle(&mut ctx, &LifeCycle::BuildFocusChain);
+    if state.item.accepts_focus {
+        state.item.focus_chain.push(id);
```

Comment:
Why don't we just extend the global focus chain here? That is, what does the per-widget focus chain achive, now that we have an explicit state for it?

Reply:
It does some caching.

It's not ideal, though. I think the whole "focus chain" concept should be refactored.

Reply:
Oh of course, some children won't actually have this pass run on them?

Reply:
Yeah, the pass doesn't run on the entire tree.